api/bucketsd/service.go

```diff
@@ -961,6 +971,247 @@ func (s *Service) createLinks(
 	}, nil
 }
 
+// MovePath moves source path to destination path and cleans up afterward
+func (s *Service) MovePath(ctx context.Context, req *pb.MovePathRequest) (res *pb.MovePathResponse, err error) {
+	log.Debugf("received move path request")
+
+	dbID, ok := common.ThreadIDFromContext(ctx)
+	if !ok {
+		return nil, errDBRequired
+	}
+	dbToken, _ := thread.TokenFromContext(ctx)
+
+	fromPth, err := parsePath(req.FromPath)
+	if err != nil {
+		return nil, err
+	}
+
+	toPth, err := parsePath(req.ToPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: enable move of root directory
+	// Can't move root directory
+	if fromPth == "" {
+		return nil, fmt.Errorf("source is root directory")
+	}
+
+	// Paths are the same, return
+	if fromPth == toPth {
+		return &pb.MovePathResponse{}, nil
+	}
+
+	// Lock bucket for the rest of transformations
+	lck := s.Semaphores.Get(buckLock(req.Key))
+	lck.Acquire()
+	defer lck.Release()
+
+	buck, pth, err := s.getBucketPath(ctx, dbID, req.Key, fromPth, dbToken)
+	if err != nil {
+		return nil, fmt.Errorf("getting path: %v", err)
+	}
+
+	buck.UpdatedAt = time.Now().UnixNano()
+	buck.SetMetadataAtPath(toPth, tdb.Metadata{
+		UpdatedAt: buck.UpdatedAt,
+	})
+	buck.UnsetMetadataWithPrefix(fromPth + "/")
+
+	if err = s.Buckets.Verify(ctx, dbID, buck, tdb.WithToken(dbToken)); err != nil {
+		return nil, fmt.Errorf("verifying bucket update: %v", err)
+	}
+
+	fromNodePath, err := inflateFilePath(buck, fromPth)
+	if err != nil {
+		return nil, err
+	}
+	fromItem, err := s.pathToItem(ctx, buck, fromNodePath, false)
+	if err != nil {
+		return nil, err
+	}
+
+	toNodePath, err := inflateFilePath(buck, toPth)
+	if err != nil {
+		return nil, err
+	}
+
+	toItem, err := s.pathToItem(ctx, buck, toNodePath, false)
+	if err == nil {
+		if fromItem.IsDir && !toItem.IsDir {
+			return nil, fmt.Errorf("not a directory")
+		}
+		if toItem.IsDir {
+			// from => to becomes cleanDst:
+			// "c" => "b" becomes "b/c"
+			// "a.jpg" => "b" becomes "b/a.jpg"
+			toPth = gopath.Join(toPth, fromItem.Name)
+		}
+	}
+
+	// bucket node
+	buckNode, err := s.getNodeAtPath(ctx, pth, buck.GetLinkEncryptionKey())
+	if err != nil {
+		return nil, fmt.Errorf("getting node: %v", err)
+	}
+
+	var dirPath path.Resolved
+	// Private buckets need to manage keys+encrypted nodes
+	if buck.IsPrivate() {
+		ctx, dirPath, err = s.copyNode(ctx, buck, buckNode, fromPth, toPth)
```

Comment:
Anything that touches the nodes should return a modified context that reflects the storage change.